pandas/tests/indexing/test_loc.py

```diff
@@ -1564,6 +1564,19 @@ def test_loc_getitem_slice_label_td64obj(self, start, stop, expected_slice):
         expected = ser.iloc[expected_slice]
         tm.assert_series_equal(result, expected)
 
+    def test_loc_getitem_slice_unordered_dt_index(self, frame_or_series):
+        # GH#18531
+        obj = frame_or_series(
+            [1, 2, 3],
+            index=[Timestamp("2017"), Timestamp("2019"), Timestamp("2018")],
+        )
+        with pytest.raises(KeyError, match=r"Timestamp\('2020-01-01 00:00:00'\)"):
+            obj.loc["2020":"2022"]
```

Comment:
Can you also add an example where the label falls within the range of the Index? 
Eg with the index being [2016, 2019, 2017] and then the slice "2018":"2020" or so

Reply:
This works, because of  https://github.com/pandas-dev/pandas/blob/72d4b78eb16f510b8d52d52912fe6f5471289de5/pandas/core/indexes/datetimes.py#L799
Should we remove or deprecate? Harder fix than the previous one

Reply:
hmm, i guess we could try to deprecate, @jorisvandenbossche you feel strongly here?